### ISSUE TITLE
[Android] Splash screen doesn't render with the `compatibility` renderer

### ISSUE THREAD

**Issue by m4gr3d** [godotengine/godot#92966]
### Tested versions

- Reproducible in Godot 4.3 beta 1

### System information

Windows 11, Android 14

### Issue description

Godot's splash screen does not show when loading a Godot app / game / editor on an Android device with the `compatibility` renderer:
- While launching, the Android editor's project manager (which uses the `compatibility` renderer) does not show the editor splash screen. 
- While launching a Godot app / game with a splash screen, the splash screen does not show with the `compatibility` renderer but shows when the `mobile` renderer is used.

### Steps to reproduce

- Android Editor:
  - Download and launch the Android Editor on an Android device
  - Notice the editor splash screen does not show

- Godot Android app / game:
  - On a desktop platform (Linux, MacOs or Windows), open the MRP
  - Set the renderer to `mobile`
  - Export and launch the MRP to an Android device
  - Validate that the splash screen shows properly
  - Now set the renderer to `compatibility`
  - Export and launch the MRP to an Android device
  - Validate that the splash screen no longer shows

### Minimal reproduction project (MRP)

[Kenney Starter Kit City Builder](https://godotengine.org/asset-library/asset/2174)

**Comment by clayjohn** [comment]
Is this a new issue with beta 1? Have you reproduced this on earlier dev builds?

**Comment by clayjohn** [comment]
Testing master (292e50e17e3f6e2509d3178a00204f964a907460) on a pixel4 and I get the same splash screen with the compatibility renderer and with the mobile renderer. In both cases it is wrong, but it doesn't show nothing.

![Screenshot_20240611-120423](https://github.com/godotengine/godot/assets/16521339/09b763fb-3883-46fe-8b5f-e06b4fa174db)

**Comment by m4gr3d** [comment]
> Testing master ([292e50e](https://github.com/godotengine/godot/commit/292e50e17e3f6e2509d3178a00204f964a907460)) on a pixel4 and I get the same splash screen with the compatibility renderer and with the mobile renderer. In both cases it is wrong, but it doesn't show nothing.
> 
> ![Screenshot_20240611-120423](https://private-user-images.githubusercontent.com/16521339/338714434-09b763fb-3883-46fe-8b5f-e06b4fa174db.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxNDEzODQsIm5iZiI6MTcxODE0MTA4NCwicGF0aCI6Ii8xNjUyMTMzOS8zMzg3MTQ0MzQtMDliNzYzZmItMzg4My00NmZlLThiNWYtZTA2YjRmYTE3NGRiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDIxMjQ0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmMjhhYTM1YmRmYjAxNmEyZDU0OGFhMTEwOTA2YmQ5ZjA1NTA5ZjA4M2ViZDljNjliZTA2OThmODJjODVkZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MeoRGwKk2dTFaKadfMuk3ol0lJbQYFwEwEkfbtwOJzc)

The splash screen you're seeing is being generated by the Android runtime and has several limitations. As a result of those limitations, we're reverting back to the Godot generated splash screen in https://github.com/godotengine/godot/pull/92965.

If you build from that PR, you can see the behavior described in this issue. To make it more noticeable, you should increase the splash screen duration in the project settings.

**Comment by m4gr3d** [comment]
> Is this a new issue with beta 1? Have you reproduced this on earlier dev builds?

The issue was reproducible in earlier dev builds.

**Comment by akien-mga** [comment]
If this is a regression, it would be good to pinpoint when it appeared. But if the issue isn't reproducible without #92965 as the splash screen doesn't even work, how do we know it's a regression, as opposed to something that never worked?

**Comment by m4gr3d** [comment]
> If this is a regression, it would be good to pinpoint when it appeared. But if the issue isn't reproducible without #92965 as the splash screen doesn't even work, how do we know it's a regression, as opposed to something that never worked?

The Godot generated splash screen was enabled for the Android editor even before #92965 ([the logic that sets `show_logo` to `false` didn't apply to `tools` builds](https://github.com/godotengine/godot/blob/07cf36d21c9056fb4055f020949fb90ebd795afb/main/main.cpp#L2543-L2547)). 
And the project manager always uses the `compatibility` renderer regardless of platform, which is how I noticed the regression.

**Comment by BastiaanOlij** [comment]
I had a  bit of a debugging session on this. For me it neither works on compatibility nor mobile renderer. I've only looked closely into the compatibility side.

Here the splash screen is prepared from `Main::setup2` where it will end up calling [set_boot_image](https://github.com/godotengine/godot/blob/master/main/main.cpp#L2967).

The implementation of this can be found in `drivers/gles3/rasterizer_gles3.cpp`, it nicely loads our image into a texture, gets some screen info (checked this works), and then blits the texture into our main system_fbo. Pretty straight forward.

But then in order to display it, it calls `end_viewport`, and `end_viewport` called `DisplayServer::get_singleton()->swap_buffers()`.

On all other systems this triggers a buffer swap displaying the result of the current buffer displaying that.

On android it sets a flag:
```
void DisplayServerAndroid::swap_buffers() {
	swap_buffers_flag = true;
}
```

That's it... the buffer is swapped for display, it's not being displayed...

The code on Android only works once the game is running. Once init is done, `onDrawFrame` gets triggered each frame, calls into `GodotLib.step()`, which prepares a frame, and returns whether our swap_buffers_flag was set, which triggers the GLSurfaceView to swap the buffers.

This logic is not accessed during setup.

Sadly my knowledge of Android is lacking to know how to fix this.
